File: BSIGeneralAffairDAL/DALCategory.vb
Imports System.Data
Imports System.Data.SqlClient
Imports BSIGeneralAffairBO
Imports GeneralAffairInterface

Public Class DALCategory
    Implements ICategory

    Public database As DatabaseConfiguration = New DatabaseConfiguration()
    Private dataRead As SqlDataReader
    Public cmd As SqlCommand

    Public Function create(obj As AssetCategory) As Integer Implements IBaseCRUD(Of AssetCategory).create
        Try
            Dim strSP = "[GeneralAffair].[USP_StoreCategoryAsset]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure

            cmd.Parameters.AddWithValue("@AssetCategoryName", obj.AssetCategoryName)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            If result = 1 Then
                Throw New ArgumentException("Category not created")
            End If
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function getAll() As List(Of AssetCategory) Implements IBaseCRUD(Of AssetCategory).getAll
        Dim Categories As New List(Of AssetCategory)
        Try
            Dim strSP = "[GeneralAffair].[USP_SelectAllCategoryAsset]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                While dataRead.Read
                    Dim Category As New AssetCategory
                    Category.AssetCategoryID = CInt(dataRead("AssetCategoryID"))
                    Category.AssetCategoryName = dataRead("AssetCategoryName").ToString()
                    Category.UpdatedAt = dataRead("UpdatedAt").ToString()
                    Categories.Add(Category)
                End While
            End If
            dataRead.Close()

            Return Categories
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function update(obj As AssetCategory) As Integer Implements IBaseCRUD(Of AssetCategory).update
        Try
            Dim updateCategory = GetById(obj.AssetCategoryID)
            If updateCategory Is Nothing Then
                Throw New ArgumentException("Category not found")
            End If

            Dim strSP = "[GeneralAffair].[USP_UpdateCategoryAsset]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@AssetCategoryID", obj.AssetCategoryID)
            cmd.Parameters.AddWithValue("@AssetCategoryName", obj.AssetCategoryName)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Private Function GetById(CategorytID As Integer) As Object
        Try
            cmd = New SqlCommand("SELECT * FROM [GeneralAffair].[AssetCategories] WHERE AssetCategoryID = @AssetCategoryID", database.conn)
            cmd.Parameters.AddWithValue("@AssetCategoryID", CategorytID)
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                dataRead.Read()
                Dim category As New AssetCategory
                category.AssetCategoryID = CInt(dataRead("AssetCategoryID"))
                category.AssetCategoryName = dataRead("AssetCategoryName").ToString
                category.UpdatedAt = dataRead("UpdatedAt").ToString()
                Return category
            Else
                Return Nothing
            End If
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function delete(id As Integer) As Integer Implements IBaseCRUD(Of AssetCategory).delete
        Try
            Dim category = GetById(id)
            If category Is Nothing Then
                Throw New ArgumentException("category not found")
            End If

            Dim strSP = "[GeneralAffair].[USP_DeleteCategoryAsset]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@AssetCategoryID", id)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function
End Class

File: BSIGeneralAffairDAL-C/DALCategory.cs
using BSIGeneralAffairBO_C;
using BSIGeneralAffairDAL_C.Interfaces;
using Dapper;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SqlClient;
using System.Text;

namespace BSIGeneralAffairDAL_C
{
    public class DALCategory : ICategoryDAL
    {
        public void Delete(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[GeneralAffair].[USP_DeleteCategoryAsset]";
                var param = new { AssetCategoryID = id };
                try
                {
                    int result = conn.Execute(strSql, param);
                    if (result == 1)
                    {
                        throw new ArgumentException("Delete data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }

        }

        private string GetConnectionString()
        {
            return Helper.GetConnectionString();
            //return @"Data Source=ACTUAL;Initial Catalog=LatihanDb;Integrated Security=True;TrustServerCertificate=True";
            //return ConfigurationManager.ConnectionStrings["MyDbConnectionString"].ConnectionString;
        }

        public IEnumerable<AssetCategory> GetAll()
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                //var strSql = @"select * from Categories order by CategoryName";
                var strSql = @"[GeneralAffair].[USP_SelectAllCategoryAsset]";

                var results = conn.Query<AssetCategory>(strSql, commandType: System.Data.CommandType.StoredProcedure);
                return results;
            }
        }

        public AssetCategory GetById(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[AssetCategories] Where [AssetCategoryID] = @AssetCategoryID ";
                var param = new { AssetCategoryID = id };
                var result = conn.QuerySingleOrDefault<AssetCategory>(strSql, param);
                return result;
            }
        }

        public IEnumerable<AssetCategory> GetByNamesCategory(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[AssetCategories] Where [AssetCategoryName] = @AssetCategoryName ";
                var param = new { AssetCategoryName = name };
                var result = conn.Query<AssetCategory>(strSql, param);
                return result;
            }
        }

        public int GetCountCategory(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT COUNT(*) FROM [GeneralAffair].[AssetCategories] Where [AssetCategoryName] = @AssetCategoryName ";
                var param = new { AssetCategoryName = $"%{name}%" };
                var result = Convert.ToInt32(conn.ExecuteScalar(strSql, param));
                return result;
            }
        }

        public IEnumerable<AssetCategory> GetWithPagingCategory(int pageNumber, int pageSize, string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[AssetCategories]
                              Where [AssetCategoryName] like @AssetCategoryName 
                              order by AssetCategoryName OFFSET @Offset ROWS FETCH NEXT @PageSize ROWS ONLY";
                var param = new { AssetCategoryName = $"%{name}%", Offset = (pageNumber - 1) * pageSize, PageSize = pageSize };
                var results = conn.Query<AssetCategory>(strSql, param);
                return results;
            }
        }

        public void Insert(AssetCategory entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = "[GeneralAffair].[USP_StoreCategoryAsset]";
                var param = new
                {
                    AssetCategoryName = entity.AssetCategoryName
                };
                try
                {
                    int result = conn.Execute(strSql, param, commandType: System.Data.CommandType.StoredProcedure);
                    if (result == 1)
                    {
                        throw new ArgumentException("Insert data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public void Update(AssetCategory entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                try
                {
                    var strSql = @"[GeneralAffair].[USP_UpdateCategoryAsset]";
                    var param = new
                    {
                        AssetCategoryID = entity.AssetCategoryID,
                        AssetCategoryName = entity.AssetCategoryName
                    };
                    int result = conn.Execute(strSql, param);

                    //jika result = -1, berarti update data gagal
                    if (result == 1)
                    {
                        throw new Exception("Update data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new Exception($"{sqlEx.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new Exception(ex.Message);
                }
            }
        }

    }
}
